File: Design-Patterns-Using-JavaScript/Interpreter/InterpreterPattern2.js
// Abstract Expression
class Expression {
    interpret(context) { }
}

// Terminal Expression
class NumberExpression extends Expression {
    constructor(number) {
        super();
        this.number = number;
    }

    interpret() {
        return this.number;
    }
}

// Non-terminal Expression
class AddExpression extends Expression {
    constructor(leftExpression, rightExpression) {
        super();
        this.leftExpression = leftExpression;
        this.rightExpression = rightExpression;
    }

    interpret(context) {
        return this.leftExpression.interpret(context) + this.rightExpression.interpret(context);
    }
}

// Context
class Context {
    constructor() {
        this.variables = new Map();
    }

    setVariable(variable, value) {
        this.variables.set(variable, value);
    }

    getVariable(variable) {
        return this.variables.get(variable) || 0;
    }
}

// Client code
const context = new Context();
context.setVariable("x", 10);
context.setVariable("y", 5);

// Create the expression tree: x + (y + 2)
const expression = new AddExpression(
    new NumberExpression(context.getVariable("x")),
    new AddExpression(
        new NumberExpression(context.getVariable("y")),
        new NumberExpression(2)
    )
);

const result = expression.interpret(context);
console.log("Result:", result);  // Output: Result: 17

/*
Result: 17
*/

File: Design-Patterns-Using-CSharp/Interpreter/InterpreterPattern2.cs
using System;
using System.Collections.Generic;

// Abstract Expression
interface Expression {
    int Interpret();
}

// Terminal Expression
class NumberExpression : Expression {
    private int number;

    public NumberExpression(int number) {
        this.number = number;
    }

    public int Interpret() {
        return number;
    }
}

// Non-terminal Expression
class AddExpression : Expression {
    private Expression leftExpression;
    private Expression rightExpression;

    public AddExpression(Expression leftExpression, Expression rightExpression) {
        this.leftExpression = leftExpression;
        this.rightExpression = rightExpression;
    }

    public int Interpret() {
        return leftExpression.Interpret() + rightExpression.Interpret();
    }
}

// Context
class Context {
    private Dictionary<string, int> variables = new Dictionary<string, int>();

    public void SetVariable(string variable, int value) {
        variables[variable] = value;
    }

    public int GetVariable(string variable) {
        return variables.TryGetValue(variable, out var value) ? value : 0;
    }
}

// Client code
class InterpreterPattern2 {
    static void Main(string[] args) {
        Context context = new Context();
        context.SetVariable("x", 10);
        context.SetVariable("y", 5);

        // Create the expression tree: x + (y + 2)
        Expression expression = new AddExpression(
            new NumberExpression(context.GetVariable("x")),
            new AddExpression(
                new NumberExpression(context.GetVariable("y")),
                new NumberExpression(2)
            )
        );

        int result = expression.Interpret();
        Console.WriteLine("Result: " + result);  // Output: Result: 17
    }
}

/*
Result: 17
*/